EIPS/eip-7702.md

```diff
@@ -0,0 +1,97 @@
+* **Sponsorship**: account X pays for a transaction on behalf of account Y. Account X could be paid in some other ERC-20 for this service, or it could be an application operator including the transactions of its users for free.
+* **Privilege de-escalation**: users can sign sub-keys, and give them specific permissions that are much weaker than global access to the account. For example, you could imagine a permission to spend ERC-20 tokens but not ETH, or to spend up to 1% of total balance per day, or to interact only with a specific application.
+
+[EIP-3074](./eip-3074.md) solves all of these use cases. However, it has forward-compatibility concerns:
+
+* It introduces two opcodes, `AUTH` and `AUTHCALL`, that would have no use in an "endgame account abstraction" world where eventually all users are using smart contract wallets (which seems like it must happen eventually, at the least because eventually quantum computers will break the ECDSA that EOAs use)
+* It leads to the development of an "invoker contract" ecosystem that would be separate from the "smart contract wallet" ecosystem, leading to possible fragmentation of effort.
+
+The purpose of this EIP is to enable all of the use cases of EIP-3074, without these two weaknesses.
+
+## Specification
+
+The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Parameters
+
+* `FORK_BLKNUM` = `TBD`
+* `TX_TYPE` = `TBD`
+* `MAGIC` = `TBD`
+* `PER_CONTRACT_CODE_BASE_COST` = `5000`
+
+As of `FORK_BLOCK_NUMBER`, a new [EIP-2718](./eip-2718.md) transaction is introduced with `TransactionType` = `TX_TYPE(TBD)`.
+
+The [EIP-2718](./eip-2718.md) `TransactionPayload` for this transaction is
+
+```
+rlp([chain_id, nonce, max_priority_fee_per_gas, max_fee_per_gas, gas_limit, destination, data, access_list, [[contract_code, y_parity, r, s], ...], signature_y_parity, signature_r, signature_s])
+```
+
+The intrinsic cost of the new transaction is inherited from [EIP-2930](./eip-2930.md), specifically `21000 + 16 * non-zero calldata bytes + 4 * zero calldata bytes + 1900 * access list storage key count + 2400 * access list address count`. Additionally, we add a cost of `16 * non-zero calldata bytes + 4 * zero calldata bytes` over each `contract_code`, plus `PER_CONTRACT_CODE_BASE_COST` times the length of the `contract_code` array.
+
+At the start of executing the transaction, for each `[contract_code, y_parity, r, s]` tuple:
+
+1. Let `signer = ecrecover(keccak(MAGIC + contract_code), y_parity, r, s]`.
+2. Verify that the contract code of `signer` is empty.
+3. Set the contract code of `signer` to `contract_code`.
+
+At the end of the transaction, set the `contract_code` of each `signer` back to empty.
+
```

Comment:
So I presume transaction execution the `EXTCODEHASH` will return the `contract_code`, right?